Fixes #23168 - undefined method `split' for nil:NilClass

ik5 opened the pull request:
A small refactoring for the helper method.
It fixes the bug, and also make it a bit easier to understand what is been looked up here.

Commit:
Fixes #23168 - undefined method `split' for nil:NilClass

Commit:
Fixes #23168 - undefined method `split' for nil:NilClass

ohadlevy reviewed (dismissed):
tests please? it is very hard to understand what did you change and if the existing usage cases still work?

dLobatog commented on app/helpers/application_helper.rb:
```diff
@@ -4,32 +4,33 @@ module ApplicationHelper
 
   # Add autogenerated html data-id attribute if not present (prefixed with "aid_")
   def link_to(*args, &block)
```
Could you remove `&block`? It's supposed to be faster, https://github.com/JuanitoFatas/fast-ruby/blob/master/code/proc-and-block/proc-call-vs-yield.rb

ik5 replied:
done

dLobatog reviewed (dismissed):
@ik5 How did you reproduce the error? https://projects.theforeman.org/issues/23168#note-1

ik5 commented:
@ohadlevy, it is not simple to test.
I refactored all of the code that I actually understand, but there are places where I still do not understand enough.
The helper is written with many code smells, and assume that there  are always specific indexes that provided to the method, and that are found inside.

That was the reason for the issue itself, it accessed a nil element.
So I refactored only code that I understand what it does and why, and kept everything else as-is.

Another change was to remove the `if else` block due to so many lines of code inside that `else` statement.

**tl;dr**:
I'll try to create few tests for it, but it will be far from a complete cover for it, I'm not sure that I know to cover all cases

tbrisker commented on app/helpers/application_helper.rb:
```diff
@@ -3,33 +3,34 @@ module ApplicationHelper
+        # only create it once...
+        @uri_keys ||= [ :scheme, :userinfo, :host, :port, :registry, :path, :opaque, :query, :fragment ]
+        uri_elements = Hash[@uri_keys.zip(URI.split(url_for(options) || html_options['href']))]
+        path = (uri_elements[:path] || '').split('/').select {|x| !x.empty?}
```
You can make this much simpler without relying on URI internals
```ruby
path = URI(url_for(options) || html_options['href']).path.split('/').reject(&:empty?)
```

ik5 replied:
I tried to do two big things: solve as much as possible the access to an array element using a literal index location, and the ability to try and keep things the same as possible, so only code smells that I understand what they do will be changed.

having said that, your solution is more elegant :D

tbrisker reviewed (changes requested):
Code certainly looks better now, please add a test at least for the edge case that caused the error.

ik5 commented:
created the test for the original issue.

tbrisker commented:
While the refactoring makes this code much more readable, looks like the error will still be logged just the same, since `URI('javascript:foo()').path==nil`. 
@ohadlevy - are you sure this is actually a bug? looks like the function is working as expected, generating a `not_parseable` id when a url that doesn't include a path is passed to it. Perhaps we need to also need to exempt urls that start with `javascript:`, similar to what we do for `mailto:`?

ohadlevy commented:
@tbrisker perhaps, all i did in this case was to observe an error in the logs and create an issue out of it....:-\

ik5 commented:
@tomer, there is an RFC to know the type of urn (schema), but the Ruby url module does not support it, looking to see if someone wrote an RFC based gem for it first, if not, I'll use the `&.` syntax of Ruby to use it on a non `nil` methods.

The reason to first look for it, is that it can help for other stuff at foreman that needs to understand urn or entire url.

tbrisker commented:
@ik5 wrong Tomer mentioned :) there is `URI#scheme` if we want, but i think just filtering javascript links like we do mailto links is fine to solve this

ekohl commented on app/helpers/application_helper.rb:
```diff
@@ -3,33 +3,30 @@ module ApplicationHelper
-        html_options[:data][:confirm] = html_options.delete(:confirm)
-      end
-      super(args[0], args[1], html_options)
+      html_options[:'data-id'] = "aid_#{id}"
```
Without looking at the details, should this even set `data-id` if id is `not_parseable` or `not_defined`? How useful is `aid_not_parseable`? Will anything ever query this and what happens if multiple elements are not parseable?

ik5 replied:
I'm not sure yet what is the usage of `data-id`, but I agree with you that it is a good thing to check if `id` actually present, and exists, changing that

ik5 commented:
okay, changed both the helper and tests a bit

tbrisker commented on app/helpers/application_helper.rb:
```diff
@@ -3,33 +3,30 @@ module ApplicationHelper
+
+    options = args[1] || {}
+    html_options = args[2] || {}
+    unless html_options.has_key?(:'data-id') || (options.is_a?(String) && options =~ /^(mailto|javascript):/)
```
Why not leave it as `starts_with?`? it's much more readable than a regex

ik5 replied:
It doesn't make it more readable when you start adding many options, but clatter the boolean validation here.

Imagine something like this:

`unless html_options.has_key?(:'data-id') || (options.is_a?(String) && options.start_with?('mailto:', 'javascript:', 'tel:', 'color:', 'date:', 'foo:'))`

It takes you few seconds to read it and then realize that each of it is an option that we are are looking for to see if the string start with, but actually the regex makes it simpler, because if we read it from left to right, we see the carrot sign and then a group of words that we are looking for at the beginning, and only then a colon sign.

So I think that the more options you have, `start_with?` because a bit less readable, and makes you think twice or more of the meaning, at least for me :/

tbrisker commented on test/helpers/application_helper_test.rb:
```diff
@@ -87,4 +87,11 @@ def test_generate_link_for
   end
+
+  describe 'link_to exceptions' do
+    test 'test if having a javascript as a path crashes link_to' do
```
the test name should say what is the expected result, not the wrong result - e.g. `link_to generates correct html for links to javascript`

tbrisker reviewed (changes requested):
a couple of minor comment inline

ohadlevy commented:
afair its used for testing purposes by satellite QE to ensure every element
has an id, @lzap?

On Mon, Jul 2, 2018 at 5:45 AM Ido Kanner <notifications@github.com> wrote:

> *@ik5* commented on this pull request.
> ------------------------------
>
> In app/helpers/application_helper.rb
> <https://github.com/theforeman/foreman/pull/5712#discussion_r199374861>:
>
> >        end
> -      if html_options[:confirm]
> -        html_options[:data] ||= {}
> -        html_options[:data][:confirm] = html_options.delete(:confirm)
> -      end
> -      super(args[0], args[1], html_options)
> +      html_options[:'data-id'] = "aid_#{id}"
>
> I'm not sure yet what is the usage of data-id, but I agree with you that
> it is a good thing to check if id actually present, and exists, changing
> that
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5712#discussion_r199374861>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx5HHB4ZxBKbs9Thci56wPRXU-NlJks5uCYlRgaJpZM4Us1tq>
> .
>

lzap commented:
Yes but I never got feedback if they actually using it. Let's check and if not, drop it?

ik5 commented:
@tbrisker i changed to `start_with?`, is there anything else you see that should be changed?

ik5 commented:
[test]

ik5 commented:
[test foreman]

ik5 commented:
[test katello]

ik5 commented:
@tbrisker wdy?

tbrisker commented:
Thanks @ik5 !

tbrisker commented:
@omaciel can you please confirm if the `data-id=aid...` attribute is being used for test automation or if we can get rid of it?

omaciel commented:
@tbrisker the following files have several paths relying on `@data-id`:

* https://github.com/SatelliteQE/robottelo/blob/master/robottelo/ui/locators/base.py
* https://github.com/SatelliteQE/robottelo/blob/master/robottelo/ui/locators/common.py

The good news is that our UI automation for recent (and future) versions of TheForeman won't rely on this code :)

The bad news, however, is that the new UI code has 3 locators that use it:

* https://github.com/SatelliteQE/airgun/blob/master/airgun/views/domain.py
* https://github.com/SatelliteQE/airgun/blob/master/airgun/views/puppet_environment.py

How should we proceed?

ekohl commented:
@omaciel as I read it keeping it in is no problem, but if it was unused then there'd be no point in keeping it around.

oshtaier commented:
The less redundant attributes - the better for us. I think we can definitely find alternative way to address these elements

abalakh commented:
@tbrisker @lzap Submitted https://github.com/SatelliteQE/airgun/pull/167 to drop all the locators with `@data-id` attribute from UI automation, feel free to drop this attr from foreman.
